<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Welcome file</title>
  <link rel="stylesheet" href="https://stackedit.io/style.css" />
</head>

<body class="stackedit">
  <div class="stackedit__html"><h1 id="spermatogenesis-differential-isoform-usage-analysis-by-scalpel">SPERMATOGENESIS differential isoform usage analysis by SCALPEL</h1>
<p><strong>Franz AKE<br>
PLASS lab, Barcelona</strong></p>
<p>The analysis can be achieved using the isoform counts matrix generated by SCALPEL, or by loading the Seurat objects generated by the tool. To perform the single-cell analysis, we use the popular library Seurat:</p>
<h2 id="a.-samples-reading">A. Samples reading</h2>
<pre><code>#reading SCALPEL seurat objects
SAMPLES.paths = c(
  "datas/SRR612905X_exp2/SRR6129050/SRR6129050_seurat.RDS",
  "datas/SRR612905X_exp2/SRR6129051/SRR6129051_seurat.RDS")

#Seurat objects reading
my.objs &lt;- lapply(SAMPLES.paths, function(x) readRDS(x))

#merging the different samples
my.obj &lt;- merge(my.objs[[1]], my.objs[2])
my.obj = JoinLayers(my.obj)

#Add the barcodes annotation into the meta.data slot
my.obj$Barcodes = str_extract(colnames(my.obj), pattern = "(A|T|G|C)*\\.[0-9]") %&gt;%
  str_replace(pattern = "\\.",replacement = "\\-")
head(my.obj, 5)
</code></pre>
<h2 id="b.-quality-filtering">B. Quality Filtering</h2>
<p>The quality filtering can be performed by subsetting curated barcodes from the current dataset, or discarding low quality cells directly from the current Seurat object. We subsetted the dataset based on curated list of cells from the study of Lukassen et al () and annotated the cells:</p>
<p><strong>NB:</strong> <em>SCALPEL execution can be performed directly on the curated list of cells by providing a curated barcodes file</em><br>
<strong>[-- barcodes 	 BARCODE_FILE]</strong></p>
<pre><code>#Filtering of the current object from the curated cell barcodes
BARCODES.path = "datas/LUKASSEN_curated_cells_barcodes.txt"

#a. read the barcodes file
curated.bcs = fread(BARCODES.path)

#b. Filtering
my.obj1 = subset(my.obj, Barcodes %in% curated.bcs$Barcodes)

#c. Annotation of cells
my.obj1$Cell_Clusters = dplyr::left_join(my.obj1@meta.data, curated.bcs)$Cell_Cluster %&gt;%
  as.factor()
</code></pre>
<h2 id="c.-dimensionality-reduction">C. Dimensionality reduction</h2>
<pre><code>#Dimensionality reduction
my.obj1 = NormalizeData(my.obj1)
my.obj1 = FindVariableFeatures(my.obj1, nfeatures = 3000)
my.obj1 = ScaleData(my.obj1)
my.obj1 = RunPCA(my.obj1)
#get number of principal components 
ElbowPlot(my.obj1,ndims = 50)
nb.pcs = 8
#get UMAP
my.obj1 = RunUMAP(my.obj1, dims = 1:nb.pcs)

#Visualization
DimPlot(my.obj1, group.by = "Cell_Clusters", label = T, pt.size = 1, label.size = 8) +
  theme_few()
</code></pre>
<p><img src="https://data.cyverse.org/dav-anon/iplant/home/franzx5/SPERMATOGENESIS/UMAP_clusters.png" alt="Isoform quantification based UMAP plot "></p>
<h2 id="d.-differential-isoform-analysis">D. Differential isoform analysis</h2>
<pre><code>#Differential isoform usage analysis

#source SCALPEL functions scripts
source("~/SCALPEL/src/scalpel_library.R")

#perform differential isoforms usage analysis in Cell Clusters
my.isoforms = Find_isoforms(my.obj1, pval_adjusted = 0.05, condition = "Cell_Clusters")
tail(my.isoforms)

#ex;
gene.in = "Arl2bp"
diff.tab = dplyr::filter(my.isoforms, gene==gene.in)
print(diff.tab)
</code></pre>

<table>
<thead>
<tr>
<th>ES</th>
<th>RS</th>
<th>SC</th>
<th>gene</th>
<th>p_value</th>
<th>p_value.adjusted</th>
<th>gene_tr</th>
<th>transcript</th>
</tr>
</thead>
<tbody>
<tr>
<td>113</td>
<td>5686</td>
<td>5971</td>
<td><strong>Arl2bp</strong></td>
<td>0</td>
<td>0</td>
<td><em>Arl2bp</em>*<em>Arl2bp-201</em></td>
<td><strong>Arl2bp-201</strong></td>
</tr>
<tr>
<td>5184</td>
<td>5660</td>
<td>1407</td>
<td><strong>Arl2bp</strong></td>
<td>0</td>
<td>0</td>
<td><em>Arl2bp</em>*<em>Arl2bp-201</em></td>
<td><strong>Arl2bp-201</strong></td>
</tr>
</tbody>
</table><h2 id="e.-visualizations-isoform-differential-usage">E. Visualizations (Isoform differential usage)</h2>
<h3 id="isoform-relative-expression-in-cell-types">1. Isoform relative expression in cell types</h3>
<pre><code>#The relative expression of APA genes between the conditions can be visualized
#using Built in function (plot_relativeExp):

plot_relativeExp(my.obj1, features_in = diff.tab$gene_tr, group.var = "Cell_Clusters")
</code></pre>
<p><img src="https://data.cyverse.org/dav-anon/iplant/home/franzx5/SPERMATOGENESIS/Boxplot_isofs.png" alt="Isoforms relative expression in cells"></p>
<h3 id="isoform-expression-in-cells-featureplot">2. Isoform expression in cells (FeaturePlot)</h3>
<pre><code>#Features plots
Idents(my.obj1) = my.obj1$Cell_Clusters

FeaturePlot(my.obj1, features = diff.tab$gene_tr, order = T, label = T, label.size = 9)
</code></pre>
<p><img src="https://data.cyverse.org/dav-anon/iplant/home/franzx5/SPERMATOGENESIS/UMAP_fplots.png" alt="Feature Plot"></p>
<h3 id="isoform-reads-coverage-track">3. Isoform reads coverage track</h3>
<p>The read coverage on the isoforms can be visualized accordingly to the defined clusters for each gene:</p>
<pre><code>#The reads coverage plots requires to merge the sample BAM files and split by each clusters
#We can rely on samtools in this purpose:

samtools.bin = "/usr/local/Caskroom/mambaforge/base/bin/samtools"
SRR6129050.BAM = "datas/SRR612905X_exp2/SRR6129050/SRR6129050.filtered.bam"
SRR6129051.BAM = "datas/SRR612905X_exp2/SRR6129051/SRR6129051.filtered.bam"
BC.tag = "CB" (depending of the sequencing protocol)

#a. merging of samples BAM files
merge.expression = paste0(samtools.bin, " merge -o SRR612905X.bam ", SRR6129050.BAM, " ", SRR6129051.BAM)
print(merge.expression)
# system(merge.expression)

#b. Splitting of the merged BAM file according to each clusters cell barcodes
lapply(unique(my.obj1$Cell_Clusters), function(x){
  print(x)
  
  #a.write cluster current barcodes
  dplyr::filter(my.obj1@meta.data, Cell_Clusters==x) %&gt;% dplyr::select(Barcodes) %&gt;%
    fwrite(paste0(x,".barcodes"), col.names = F, row.names = F)
  
  #b. subset merged bam file
  bam_subset.expression = paste0(
    samtools.bin, " view -b -D ", BC.tag, ":", paste0(x,".barcodes"), " SRR612905X.bam &gt; ", x, ".bam")
  
  #c. indexing spitted bam file
  index.expression = paste0(samtools.bin, " index ", x, ".bam")
  
  #execution
  print(bam_subset.expression)
  system(bam_subset.expression)
  
  print(index.expression)
  system(index.expression)
})

#c. Coverage Track visualization
GTF.path = "datas/gencode.vM21.annotation.gtf"
filtered.bamfiles = c(
  "ES" = "ES.bam",
  "RS" = "RS.bam",
  "SC" = "SC.bam"
)

#1. Create genomeRange object
gtf.gr = rtracklayer::import(GTF.path)

#2.plot
CoverPlot(genome_gr = gtf.gr, gene_in = gene.in, genome_sp = "mm10",
          transcripts_in = diff.tab$transcript, filter_trs = T,
          bamfiles = filtered.bamfiles, samtools.bin = samtools.bin)
</code></pre>
<p><img src="https://data.cyverse.org/dav-anon/iplant/home/franzx5/SPERMATOGENESIS/Trackplot.png" alt="enter image description here"></p>
<h2 id="f.-3utr-size-dynamic-in-cell-types">F. 3’UTR size dynamic in cell types</h2>
<p>Using to the isoform quantification, we can weight the quantified isoform 3’UTR length by their expression and analyze the average 3’UTR size dynamic in the cell types:</p>
<pre><code>#A. At single-cell resolution

#1. Get Isoform expression in cells
#**********************************
cell.exp = my.obj1[["RNA"]]$counts %&gt;% data.frame()
gene_trs = str_split_fixed(rownames(cell.exp), pattern = "\\*\\*\\*", 2)
cell.exp = cell.exp %&gt;%
  mutate(gene_transcript=rownames(cell.exp), gene_name=gene_trs[,1], transcript_name=gene_trs[,2]) %&gt;%
  melt() %&gt;%
  dplyr::filter(value&gt;0)
#annotate with cell types
cell.exp = cell.exp %&gt;%
  mutate(variable = str_replace(variable,"\\.","-") %&gt;% str_extract(pattern = "(A|T|G|C)*\\-[0-9]")) %&gt;% 
  left_join(my.obj1@meta.data[,c("Barcodes","Cell_Clusters")], by = join_by(variable==Barcodes),
            relationship = "many-to-many") %&gt;%
  data.table()

#2. Get 3'UTR size
#*****************
utrTab = data.frame(gtf.gr) %&gt;%
  filter(transcript_name %in% cell.exp$transcript_name) %&gt;%
  dplyr::filter(type %in% c("UTR","CDS")) %&gt;%
  mutate(utr_type=NA) %&gt;%
  group_by(gene_name, transcript_name, type) %&gt;%
  #get coding region (CDS) coordinates
  mutate(cds_start = ifelse(type=="CDS", min(start),NA), cds_end = ifelse(type=="CDS",max(end),NA)) %&gt;%
  group_by(gene_name,transcript_name) %&gt;%
  mutate(cds_start = cds_start[1], cds_end = cds_end[1]) %&gt;%
  filter(type=="UTR") %&gt;%
  #tag 3'UTR/5'UTR regions
  mutate(utr_type = ifelse(strand=="+" &amp; end &lt; cds_start, "5UTR","3UTR"),
         utr_type = ifelse(strand=="-" &amp; end &lt; cds_start, "3UTR","5UTR")) %&gt;%
  #subset 3'UTR regions
  dplyr::filter(utr_type=="3UTR") %&gt;%
  mutate(UTRlength = sum(width)) %&gt;%
  ungroup() %&gt;%
  distinct(transcript_name, UTRlength)

#3. Annotate Isoform with 3'UTR length
#*************************************
cell.exp2 = left_join(cell.exp, utrTab) %&gt;%
  na.omit()

#4. Weighting
#************
#*by gene
cell.exp3 = cell.exp2 %&gt;%
  group_by(Cell_Clusters, variable, gene_name) %&gt;%
  mutate(value.norm = value / sum(value), weighted.UTRlength.iso = value.norm * UTRlength,
         weighted.UTRlength.gene = mean(weighted.UTRlength.iso)) %&gt;%
  data.table()

#by Cell
cell.exp4 = distinct(cell.exp3, Cell_Clusters, variable, gene_name, weighted.UTRlength.gene) %&gt;%
  group_by(Cell_Clusters, variable) %&gt;%
  summarise(weighted.UTRlength.cell = mean(weighted.UTRlength.gene)) %&gt;%
  data.table()
  
#Plot  
ggplot(cell.exp4, aes(Cell_Clusters, weighted.UTRlength.cell, fill=Cell_Clusters)) +
  geom_boxplot(width=0.3) +
  theme_few(base_size = 13) +
  scale_fill_canva() +
  ggtitle("average weigthed gene UTR length in cells for each cluster")
</code></pre>
<p><img src="https://data.cyverse.org/dav-anon/iplant/home/franzx5/SPERMATOGENESIS/UTR_dyn.png" alt="enter image description here"></p>
</div>
</body>

</html>
